File: main/engine/v6/src/tt.cpp
#include "tt.hpp"
#include <cstring>
#include <new>

#ifdef _MSC_VER
#include <malloc.h>  // For _aligned_malloc/_aligned_free on MSVC
#endif

namespace v6 {

// Global TT instance (64MB default)
TranspositionTable TT(64);

TranspositionTable::TranspositionTable(size_t size_mb) {
    size_t size_bytes = size_mb * 1024 * 1024;
    num_entries = size_bytes / sizeof(TTEntry);
    
    // Align to cache line (64 bytes)
#ifdef _MSC_VER
    entries = static_cast<TTEntry*>(_aligned_malloc(num_entries * sizeof(TTEntry), 64));
#else
    entries = static_cast<TTEntry*>(std::aligned_alloc(64, num_entries * sizeof(TTEntry)));
#endif
    if (!entries) {
        throw std::bad_alloc();
    }
    
    clear();
}

TranspositionTable::~TranspositionTable() {
#ifdef _MSC_VER
    _aligned_free(entries);
#else
    std::free(entries);
#endif
}

void TranspositionTable::clear() {
    std::memset(entries, 0, num_entries * sizeof(TTEntry));
    hit_count = 0;
    miss_count = 0;
}

bool TranspositionTable::probe(uint64_t hash, TTEntry& entry) const {
    size_t idx = index(hash);
    const TTEntry& stored = entries[idx];
    
    if (stored.key == hash && stored.flag != TT_NONE) {
        entry = stored;
        hit_count.fetch_add(1, std::memory_order_relaxed);
        return true;
    }
    
    miss_count.fetch_add(1, std::memory_order_relaxed);
    return false;
}

void TranspositionTable::store(uint64_t hash, Move best_move, int score, int depth, TTFlag flag) {
    size_t idx = index(hash);
    TTEntry& stored = entries[idx];
    
    // Replacement strategy: always replace if:
    // 1. Same position (same hash)
    // 2. New entry has higher depth
    // 3. Old entry is from previous generation
    bool should_replace = (stored.key == hash) ||
                          (stored.depth < depth) ||
                          (stored.age != generation);
    
    if (should_replace) {
        stored.key = hash;
        stored.best_move = best_move;
        stored.score = static_cast<int16_t>(score);
        stored.depth = static_cast<int8_t>(depth);
        stored.flag = static_cast<uint8_t>(flag);
        stored.age = generation;
    }
}

} // namespace v6

File: main/engine/v5c/tt.py
"""
V4d Transposition Table - Lockless Hash Table for Move Caching
Following Chess Programming Wiki (CPW) specifications.

The transposition table stores previously searched positions to avoid
redundant work. Uses Zobrist keys for position identification.
"""

import numpy as np
from numba import njit, uint64, int32, int16
from typing import NamedTuple, Optional

# =============================================================================
# TT ENTRY FLAGS
# =============================================================================

TT_EXACT = 0    # Exact score
TT_ALPHA = 1    # Upper bound (failed low)
TT_BETA = 2     # Lower bound (failed high)

# =============================================================================
# TT ENTRY STRUCTURE
# =============================================================================

# Structure: 16 bytes per entry
# - hash: 8 bytes (uint64) - Zobrist key
# - move: 4 bytes (int32) - Best move
# - score: 2 bytes (int16) - Evaluation score
# - depth: 1 byte (int8) - Search depth
# - flag: 1 byte (int8) - Entry type (exact/alpha/beta)


class TTEntry:
    """Transposition table entry."""
    __slots__ = ['hash', 'move', 'score', 'depth', 'flag', 'age']
    
    def __init__(self, hash_key=0, move=0, score=0, depth=0, flag=TT_EXACT, age=0):
        self.hash = hash_key
        self.move = move
        self.score = score
        self.depth = depth
        self.flag = flag
        self.age = age


# =============================================================================
# TRANSPOSITION TABLE
# =============================================================================

class TranspositionTable:
    """
    Lockless transposition table using numpy arrays.
    
    Uses a simple replacement scheme: always replace if:
    - New entry is from deeper search
    - Existing entry is from an older search
    """
    
    def __init__(self, size_mb: int = 64):
        """Initialize TT with given size in megabytes."""
        # Each entry is roughly 24 bytes (with Python overhead, we estimate 32)
        # For arrays, we use separate arrays which is more cache-efficient
        self.entry_size = 32
        self.num_entries = (size_mb * 1024 * 1024) // self.entry_size
        
        # Use power of 2 for fast modulo via bitwise AND
        self.num_entries = 1 << (self.num_entries.bit_length() - 1)
        self.mask = self.num_entries - 1
        
        # Separate arrays for better cache performance
        self.keys = np.zeros(self.num_entries, dtype=np.uint64)
        self.moves = np.zeros(self.num_entries, dtype=np.int32)
        self.scores = np.zeros(self.num_entries, dtype=np.int16)
        self.depths = np.zeros(self.num_entries, dtype=np.int8)
        self.flags = np.zeros(self.num_entries, dtype=np.int8)
        self.ages = np.zeros(self.num_entries, dtype=np.uint8)  # 0-255 range
        
        self.current_age = 0
        self.hits = 0
        self.stores = 0
        
        print(f"[V4d] TT initialized: {self.num_entries} entries ({size_mb}MB)")
    
    def clear(self):
        """Clear all entries."""
        self.keys.fill(0)
        self.moves.fill(0)
        self.scores.fill(0)
        self.depths.fill(0)
        self.flags.fill(0)
        self.ages.fill(0)
        self.hits = 0
        self.stores = 0
    
    def new_search(self):
        """Called at the start of each new search."""
        self.current_age = (self.current_age + 1) % 256
    
    def probe(self, key: int) -> Optional[TTEntry]:
        """
        Look up a position in the TT.
        Returns TTEntry if found, None otherwise.
        """
        index = key & self.mask
        
        if self.keys[index] == key:
            self.hits += 1
            return TTEntry(
                hash_key=self.keys[index],
                move=int(self.moves[index]),
                score=int(self.scores[index]),
                depth=int(self.depths[index]),
                flag=int(self.flags[index]),
                age=int(self.ages[index])
            )
        
        return None
    
    def store(self, key: int, move: int, score: int, depth: int, flag: int):
        """
        Store a position in the TT.
        Uses replacement scheme: replace if deeper or older entry.
        """
        index = key & self.mask
        
        # Replacement logic: always replace if
        # 1. Empty entry
        # 2. Same position (update)
        # 3. New search has deeper depth
        # 4. Old entry is from previous search
        should_replace = (
            self.keys[index] == 0 or
            self.keys[index] == key or
            depth >= self.depths[index] or
            self.ages[index] != self.current_age
        )
        
        if should_replace:
            self.keys[index] = key
            self.moves[index] = move
            self.scores[index] = max(-32000, min(32000, score))  # Clamp
            self.depths[index] = depth
            self.flags[index] = flag
            self.ages[index] = np.uint8(self.current_age & 0xFF)
            self.stores += 1
    
    def hashfull(self) -> int:
        """Return permille of TT entries used (for UCI info)."""
        # Sample 1000 entries
        sample_size = min(1000, self.num_entries)
        used = np.count_nonzero(self.keys[:sample_size])
        return (used * 1000) // sample_size


# Global TT instance
TT = TranspositionTable(size_mb=64)
